### PR TITLE
Remove non-generic BindingDiagnosticBag

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Multi-stage work to fix https://github.com/dotnet/roslyn/issues/67785.

This is part of to remove the heap allocations for `CSharp.BindingDiagnosticBag` and `VisualBasic.BindingDiagnosticBag` by making the core `BindingDiagnosticBag<TAssemblySymbol>` a struct.  The general idea is:

1. core type becomes a struct.
2. no more subtypes.  Instead, the functionality added in teh subtypes (which themselves add no state) can be provided at those levels as extensions on the struct.
3. `BindingDiagnosticBag<TAssemblySymbol>` is merged with `CodeAnalysis.BindingDiagnosticBag`, and no one passes `CodeAnalsysis.BindingDiagnosticBag` around itself.  Thsi is needed as both those core types have their own state, and so converting up then down the psuedo-inheritance chain would not work as it would result in lost data.

Breaking into smaller PRs to make things easier to keep track of.  

Full set of PRs are:

- [ ] https://github.com/dotnet/roslyn/pull/68041 (this PR)
- [ ] https://github.com/dotnet/roslyn/pull/68042
- [ ] https://github.com/dotnet/roslyn/pull/68043
- [ ] https://github.com/dotnet/roslyn/pull/68044
- [ ] https://github.com/dotnet/roslyn/pull/68046

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs [hunk 1]
```diff
@@ -232,7 +232,7 @@ internal bool IsSecurityAttribute(CSharpCompilation compilation)
 
         #region Attribute Decoding
 
-        internal void DecodeSecurityAttribute<T>(Symbol targetSymbol, CSharpCompilation compilation, ref DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        internal void DecodeSecurityAttribute<T>(Symbol targetSymbol, CSharpCompilation compilation, ref DecodeWellKnownAttributeArguments<AssemblySymbol, AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs:235]: note: a using alias for this type would make a lot of this simpler (see later PRs for an example of this).  However, i wanted to do the minimal amount of effort here to keep the review minimal.

**Review by jjonescz — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @jcouv ptal.

**Comment by AlekseyTs** [comment]
I do not think we should be going with this change alone into the main branch. We have a number of feature branches active at the moment, and changes around BindingDiagnosticBag are going to create quite a lot of noise and distraction when the code flows into those branches, especially in multiple waves. If I understand correctly, our ultimate goal is to make BindingDiagnosticBag a struct, this is definitely something worth trying. However, I would object to merging intermediate steps into the main. My preference is merging the final result, when the ultimate goal is achieved and we are comfortable with the result. There are at least two options to get there:
- Put all work in one PR
- Create a feature branch for this refactoring. Complete work in that feature branch. Merge into main at the end in one go.

**Comment by CyrusNajmabadi** [comment]
> Create a feature branch for this refactoring. Complete work in that feature branch. Merge into main at the end in one go.

Of the two options, this would be my preference, as otherwise a single PR will end up with a lot of code/concepts to review.

**Comment by CyrusNajmabadi** [comment]
ok, changed to go into new branch "features/StructDiagnosticBag". @dotnet/roslyn-compiler for second review plz.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  for second pair of eyes plz for this feature branch pr.  Thanks!

**Review by cston — APPROVED** [review]

**Pull request merged**